extensions/http-curl/client/HTTPCallback.h

```diff
@@ -33,148 +34,195 @@ namespace minifi {
+          ptr,
+          current_vec_.size(),
+          current_buffer_start_,
+          current_pos_,
+          total_bytes_loaded_);
+    }
   }
 
-  virtual const size_t getRemaining(size_t pos) {
-    return current_vec_.size();
-  }
+  /**
+   * Common implementation for placing a buffer into the queue
+   * @param vec the buffer to be inserted
+   * @return the number of bytes processed (the size of vec)
+   */
+  int64_t processInner(std::vector<char>&& vec) {
+    size_t size = vec.size();
 
-  virtual const size_t getBufferSize() {
-    std::lock_guard<std::recursive_mutex> lock(mutex_);
+    logger_->log_trace("processInner() called, vec.data(): %p, vec.size(): %zu", vec.data(), size);
 
-    if (ptr == nullptr || current_pos_ >= rolling_count_) {
-      load_buffer();
+    if (size == 0U) {
+      return 0U;
     }
-    return rolling_count_;
-  }
 
- private:
+    std::unique_lock<std::mutex> lock(mutex_);
+    byte_arrays_.emplace_back(std::move(vec));
+    cv.notify_all();
 
-  inline void load_buffer() {
-    std::unique_lock<std::recursive_mutex> lock(mutex_);
-    cv.wait(lock, [&] {return byte_arrays_.size_approx() > 0 || is_alive_==false;});
-    if (!is_alive_ && byte_arrays_.size_approx() == 0) {
-      lock.unlock();
-      return;
+    return size;
+  }
+
+  /**
+   * Seeks to the specified position
+   * @param lock unique_lock which *must* own the lock
+   * @param pos position to seek to
+   */
+  void seekInner(std::unique_lock<std::mutex>& lock, size_t pos) {
+    logger_->log_trace("seekInner() called, current_pos_: %zu, pos: %zu", current_pos_, pos);
+    if (pos < current_pos_) {
+      const std::string errstr = "Seeking backwards is not supported, tried to seek from " + std::to_string(current_pos_) + " to " + std::to_string(pos);
+      logger_->log_error("%s", errstr);
+      throw std::logic_error(errstr);
     }
-    try {
-      if (byte_arrays_.try_dequeue(current_vec_)) {
-        ptr = &current_vec_[0];
-        previous_pos_.store(rolling_count_.load());
-        current_pos_ = 0;
-        rolling_count_ += current_vec_.size();
-      } else {
-        ptr = nullptr;
+    while ((pos - current_buffer_start_) >= current_vec_.size()) {
+      loadNextBuffer(lock);
+      if (ptr == nullptr) {
+        break;
       }
-      lock.unlock();
-    } catch (...) {
-      lock.unlock();
     }
   }
 
-  std::atomic<bool> is_alive_;
-  std::atomic<size_t> rolling_count_;
-  std::condition_variable_any cv;
-  std::atomic<size_t> previous_pos_;
-  std::atomic<size_t> current_pos_;
+  std::shared_ptr<logging::Logger> logger_;
 
-  std::recursive_mutex mutex_;
+  std::mutex mutex_;
+  std::condition_variable cv;
 
-  moodycamel::ConcurrentQueue<std::vector<char>> byte_arrays_;
+  bool is_alive_;
+  size_t total_bytes_loaded_;
+  size_t current_buffer_start_;
+  size_t current_pos_;
 
-  char *ptr;
+  std::deque<std::vector<char>> byte_arrays_;
 
   std::vector<char> current_vec_;
+  char *ptr;
```

Comment:
For consistency, can you rename this to end with underscore?
I was searching the declaration of this in a function for 2 minutes :)

Reply:
Yeah, this was originally this way, I just reordered stuff, but I will change it.